### PR TITLE
protect: prompt if the user has asked the protecting admin about the …

### PR DESCRIPTION
**Pull request by MusikAnimal**
…protection when requesting unprotection

Prevent tracking down logs of circular page moves, some refactoring

Record if there was a stable log even if it's not currently pc protected

### CODE REVIEW

**Comment by MusikAnimal** [comment]
@Amorymeltzer Alright here you are :) Sorry it took me a month to respond to you, and moreover for that major major fuckup of force-pushing back to origin/master (which I've since fixed)!

File: modules/twinkleprotect.js [hunk 1]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+				if (!adminEditDeferred) {
+					adminEditDeferred = Twinkle.protect.fetchProtectingAdmin(api, mw.config.get('wgPageName'));
+				}
 			}
 		});
 
+		Twinkle.protect.hasStableLog = !!stableData[0].query.logevents.length;
+
 		if (page.flagged) {
+			// note that stable settings aren't logged when page is moved, so we don't need to use fetchProtectingAdmin
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]: This seems to longer be the case, being taken care of in [T59912](https://phabricator.wikimedia.org/T59912) a year after your original pull, so we'll have to redo this bit.

Comment (Amorymeltzer) [reply modules/twinkleprotect.js]: I think it should still be doable with `stableData`, is that right?  If we check through `stableData[0].query.logevents` for the first entry without `"action": "move_stable"`.  Annoyingly, `leaction` seems to only allow one parameter (`stable/config` or `stable/modify` but not both) and doesn't allow negative filtering (can't exclude `stable/move_stable` or `protect/move_prot`) which would make this whole thing easier.

File: modules/twinkleprotect.js [hunk 2]
```diff
@@ -98,6 +98,31 @@ Twinkle.protect.callback = function twinkleprotectCallback() {
 // { edit: { level: "sysop", expiry: <some date>, cascade: true }, ... }
 Twinkle.protect.currentProtectionLevels = {};
 
+// returns a jQuery Deferred object, usage:
+//   Twinkle.protect.fetchProtectingAdmin(apiObject, pageName).done(function(admin_username) { ...code... });
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
//   Twinkle.protect.fetchProtectingAdmin(apiObject, pageName, protect/stable).done(function(admin_username) { ...code... });
```

File: modules/twinkleprotect.js [hunk 3]
```diff
@@ -98,6 +98,31 @@ Twinkle.protect.callback = function twinkleprotectCallback() {
...
+Twinkle.protect.fetchProtectingAdmin = function twinkleprotectFetchProtectingAdmin(api, pageName, logIds) {
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
Twinkle.protect.fetchProtectingAdmin = function twinkleprotectFetchProtectingAdmin(api, pageName, protType, logIds) {
```

File: modules/twinkleprotect.js [hunk 4]
```diff
@@ -98,6 +98,31 @@ Twinkle.protect.callback = function twinkleprotectCallback() {
...
+	logIds = logIds || [];
+
+	return api.get({
+		format: 'json',
+		action: 'query',
+		list: 'logevents',
+		letitle: pageName,
+		letype: 'protect'
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
		letype: protType
```

File: modules/twinkleprotect.js [hunk 5]
```diff
@@ -98,6 +98,31 @@ Twinkle.protect.callback = function twinkleprotectCallback() {
...
+	}).then(function( data ) {
+		// don't check log entries that have already been checked (e.g. don't go into an infinite loop!)
+		var event = data.query ? $.grep(data.query.logevents, function(le) { return $.inArray(le.logid, logIds); })[0] : null;
+		if (!event) {
+			// fail gracefully
+			return null;
+		} else if (event.action === "move_prot") {
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
		} else if (event.action === "move_prot" || event.action === "move_stable") {
```

File: modules/twinkleprotect.js [hunk 6]
```diff
@@ -98,6 +98,31 @@ Twinkle.protect.callback = function twinkleprotectCallback() {
...
+			return twinkleprotectFetchProtectingAdmin( api, event.params.oldtitle_title, logIds.concat(event.logid) );
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
			return twinkleprotectFetchProtectingAdmin( api, (protType === 'protect' ? event.params.oldtitle_title : event.params.oldtitle), protType, logIds.concat(event.logid) );
```

So damn annoying.

File: modules/twinkleprotect.js [hunk 7]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
 					expiry: protection.expiry,
 					cascade: protection.cascade === ''
 				};
+				// logs report last admin who made changes to either edit/move/create protection, regardless if they only modified one of them
+				if (!adminEditDeferred) {
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]: I couldn't quite figure out why this was necessary?  It seems smart, but I couldn't trigger any errors when removing it...

File: modules/twinkleprotect.js [hunk 8]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+					adminEditDeferred = Twinkle.protect.fetchProtectingAdmin(api, mw.config.get('wgPageName'));
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
					adminEditDeferred = Twinkle.protect.fetchProtectingAdmin(api, mw.config.get('wgPageName'), 'protect');
```

File: modules/twinkleprotect.js (hunk 1 above)
Comment (Amorymeltzer) [reply modules/twinkleprotect.js]:
```suggestion
```

File: modules/twinkleprotect.js [hunk 9]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+				}
 			}
 		});
 
+		Twinkle.protect.hasStableLog = !!stableData[0].query.logevents.length;
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
```
Move back down, not needed up here anymore

File: modules/twinkleprotect.js [hunk 10]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
```

File: modules/twinkleprotect.js [hunk 11]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
 		if (page.flagged) {
+			// note that stable settings aren't logged when page is moved, so we don't need to use fetchProtectingAdmin
 			current.stabilize = {
 				level: page.flagged.protection_level,
-				expiry: page.flagged.protection_expiry
+				expiry: page.flagged.protection_expiry,
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
				expiry: page.flagged.protection_expiry
```

File: modules/twinkleprotect.js [hunk 12]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+				admin: Twinkle.protect.hasStableLog ? stableData[0].query.logevents[0].user : null
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
```

File: modules/twinkleprotect.js [hunk 13]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
 			};
 		}
 
 		// show the protection level and log info
 		Twinkle.protect.hasProtectLog = !!protectData[0].query.logevents.length;
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
		Twinkle.protect.hasProtectLog = !!protectData[0].query.logevents.length;
		Twinkle.protect.hasStableLog = !!stableData[0].query.logevents.length;
```

File: modules/twinkleprotect.js [hunk 14]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
-		Twinkle.protect.hasStableLog = !!stableData[0].query.logevents.length;
 		Twinkle.protect.currentProtectionLevels = current;
-		Twinkle.protect.callback.showLogAndCurrentProtectInfo();
+
+		if (adminEditDeferred) {
+			adminEditDeferred.done(function(admin) {
+				if (admin) {
+					$.each(['edit', 'move', 'create'], function(i, type) {
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
					$.each(['edit', 'move', 'create', 'stabilize'], function(i, type) {
```

File: modules/twinkleprotect.js [hunk 15]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+						if (Twinkle.protect.currentProtectionLevels[type]) {
+							Twinkle.protect.currentProtectionLevels[type].admin = admin;
+						}
+					});
+				}
+				Twinkle.protect.callback.showLogAndCurrentProtectInfo();
+			});
+		} else {
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
```

Unnecessary given that pc is now handled with `adminEditDeferred`

File: modules/twinkleprotect.js [hunk 16]
```diff
@@ -132,21 +157,42 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
...
+			Twinkle.protect.callback.showLogAndCurrentProtectInfo();
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]:
```suggestion
```

File: modules/twinkleprotect.js [hunk 17]
```diff
@@ -1159,7 +1214,11 @@ Twinkle.protect.callback.evaluate = function twinkleprotectCallbackEvaluate(e) {
 					typename = 'create protection';
 					break;
 				case 'unprotect':
-					/* falls through */
+					var admins = $.map(Twinkle.protect.currentProtectionLevels, function(pl) { return pl.admin ? 'User:' + pl.admin : null; });
+					if (admins.length && !confirm('Have you attempted to contact the protecting admins (' + $.unique(admins).join(', ') + ') first?' )) {
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]: Thoughts about doing a whole thing here for admin vs admins, but most of the good functions are in ES6 and it's just not really worth it.

File: modules/twinkleprotect.js (hunk 15 above)
Comment (Amorymeltzer) [reply modules/twinkleprotect.js]: Needed if not protected to show previous log, duh.

File: modules/twinkleprotect.js (hunk 16 above)
Comment (Amorymeltzer) [reply modules/twinkleprotect.js]: Needed if not protected to show previous log, duh.

File: modules/twinkleprotect.js [hunk 18]
```diff
@@ -132,6 +157,10 @@ Twinkle.protect.fetchProtectionLevel = function twinkleprotectFetchProtectionLev
 					expiry: protection.expiry,
 					cascade: protection.cascade === ''
 				};
+				// logs report last admin who made changes to either edit/move/create protection, regardless if they only modified one of them
```
Comment (Amorymeltzer) [inline modules/twinkleprotect.js]: I look forward to getting a bug report about this!

**Pull request closed without merging**